src/cc/libbpf.c

```diff
@@ -856,47 +856,100 @@ static int bpf_attach_tracing_event(int progfd, const char *event_path, int pid,
+  return 0;
+}
+
 int bpf_attach_kprobe(int progfd, enum bpf_probe_attach_type attach_type,
-                      const char *ev_name, const char *fn_name, uint64_t fn_offset)
+                      const char *ev_name, const char *fn_name, uint64_t fn_offset,
+                      int maxactive)
 {
   int kfd, pfd = -1;
-  char buf[256];
-  char event_alias[128];
-  static char *event_type = "kprobe";
+  char buf[256], fname[256];
+
+  if (maxactive <= 0)
+    // Try create the kprobe Perf Event with perf_event_open API.
+    pfd = bpf_try_perf_event_open_with_probe(fn_name, fn_offset, -1, "kprobe",
+                                             attach_type != BPF_PROBE_ENTRY);
 
-  // Try create the kprobe Perf Event with perf_event_open API.
-  pfd = bpf_try_perf_event_open_with_probe(fn_name, fn_offset, -1, event_type,
-                                           attach_type != BPF_PROBE_ENTRY);
   // If failed, most likely Kernel doesn't support the new perf_event_open API
   // yet. Try create the event using debugfs.
   if (pfd < 0) {
-    snprintf(buf, sizeof(buf), "/sys/kernel/debug/tracing/%s_events", event_type);
-    kfd = open(buf, O_WRONLY | O_APPEND, 0);
-    if (kfd < 0) {
-      fprintf(stderr, "open(%s): %s\n", buf, strerror(errno));
+    if (create_kprobe_event(buf, ev_name, attach_type, fn_name, fn_offset,
+                            maxactive) < 0)
       goto error;
-    }
-
-    snprintf(event_alias, sizeof(event_alias), "%s_bcc_%d", ev_name, getpid());
 
-    if (fn_offset > 0 && attach_type == BPF_PROBE_ENTRY)
-      snprintf(buf, sizeof(buf), "p:%ss/%s %s+%"PRIu64,
-               event_type, event_alias, fn_name, fn_offset);
-    else
-      snprintf(buf, sizeof(buf), "%c:%ss/%s %s",
-               attach_type == BPF_PROBE_ENTRY ? 'p' : 'r',
-               event_type, event_alias, fn_name);
-
-    if (write(kfd, buf, strlen(buf)) < 0) {
-      if (errno == ENOENT)
-         fprintf(stderr, "cannot attach kprobe, probe entry may not exist\n");
-      else
-         fprintf(stderr, "cannot attach kprobe, %s\n", strerror(errno));
-      close(kfd);
-      goto error;
+    // If we're using maxactive, we need to check that the event was created
+    // under the expected name.  If debugfs doesn't support maxactive yet
+    // (kernel < 4.12), the event is created under a different name; we need to
+    // delete that event and start again without maxactive.
+    if (maxactive > 0 && attach_type == BPF_PROBE_RETURN) {
+      snprintf(fname, sizeof(fname), "%s/id", buf);
+      if (access(fname, F_OK) == -1) {
+        // Deleting kprobe event with incorrect name.
+        kfd = open("/sys/kernel/debug/tracing/kprobe_events",
+                   O_WRONLY | O_APPEND, 0);
+        if (kfd < 0) {
+          fprintf(stderr, "open(/sys/kernel/debug/tracing/kprobe_events): %s\n",
+                  strerror(errno));
+          return -1;
+        }
+        snprintf(fname, sizeof(fname), "-:kprobes/%s_0", ev_name);
```

Comment:
Is it possible to re-use detach kprobe logic here?

Reply:
I considered it, but `bpf_detach_probe` also contains some code to handle the case where we attached the probe using `perf_event_open` (it iterates on the list of events). We could break `bpf_detach_probe` down to refactor, but I want to try and add support for `maxactive` in `perf_event_open`, so I thought I'd refactor after that.

Reply:
Sorry to bug a closed issue.

But where is the documentation for maxactive as referred in "See the kprobes documentation for its default value."?

We run into issue similar to https://github.com/iovisor/bcc/issues/1920 where bpf_trace_printk() states that the perf_submit() succeeded, but user-space did not read the events from the perf buffer.

Reply:
The default value is discussed in [the kprobes documentation in the kernel](https://www.kernel.org/doc/Documentation/kprobes.txt):

> If maxactive <= 0, it is
set to a default value.  If CONFIG_PREEMPT is enabled, the default
is max(10, 2*NR_CPUS).  Otherwise, the default is NR_CPUS.